packages/qwik-city/runtime/src/qwik-image-component.tsx

```diff
@@ -0,0 +1,291 @@
+    return [
+      // Always include the image at 1x and 2x the specified width
+      width,
+      doubleWidth,
+      // Filter out any resolutions that are larger than the double-res image
+      ...resolutions.filter((w) => w < doubleWidth),
+    ];
+  }
+
+  return [];
+};
+
+export const getSrcSet = ({
+  src,
+  width,
+  height,
+  aspectRatio,
+  resolutions,
+  layout,
+}: Pick<
+  QwikImageProps,
+  'src' | 'width' | 'height' | 'aspectRatio' | 'resolutions' | 'layout'
+>): string => {
+  const breakpoints = getBreakpoints({
+    width: typeof width === 'string' ? parseInt(width, 10) : width,
+    layout,
+    resolutions: resolutions || DEFAULT_RESOLUTIONS,
+  });
+  return breakpoints
+    .sort()
+    .map((bp) => {
+      let transformedHeight;
+      if (height && aspectRatio) {
+        transformedHeight = Math.round(bp * aspectRatio);
+      }
+
+      const transformed = {
+        url: src,
+        width: bp,
+        height: transformedHeight,
+      };
+
+      if (transformed) {
+        return `${transformed.url}?w=${bp} ${bp}w`;
+      }
+      return '';
+    })
+    .join(',\n');
+};
+
+type QwikImageAttributes = QwikIntrinsicElements['img'];
+
+/**
+ * @alpha
+ */
+export type BlurUrlLoaderProps = { src: string; width: number; height: number; quality: number };
+
+/**
+ * @alpha
+ */
+export type Layout = 'fixed' | 'constrained' | 'fullWidth';
+
+/**
+ * @alpha
+ */
+export type QwikImageState = {
+  blur?: boolean;
+  blurUrlLoader$?: QRL<(params: BlurUrlLoaderProps) => string>;
+  resolutions?: number[];
+};
+
+/**
+ * @alpha
+ */
+export interface QwikImageProps extends QwikImageAttributes {
+  blur?: boolean;
+  blurUrlLoader$?: QRL<(params: BlurUrlLoaderProps) => string>;
+  style?: Record<string, string | number>;
+  aspectRatio?: number;
+  layout: Layout;
+  objectFit?: 'contain' | 'cover' | 'fill' | 'none' | 'scale-down' | 'inherit' | 'initial';
+  resolutions?: number[];
+}
+
+export const QwikImageContext = createContextId<QwikImageState>('qwik-image-context');
+
+/**
+ * @alpha
+ */
+export const useImageProvider = (state: QwikImageState) => {
+  useContextProvider(QwikImageContext, state);
+};
+
+export const getStyles = ({
+  width,
+  height,
+  aspectRatio,
+  objectFit = 'cover',
+  layout,
+}: Pick<QwikImageProps, 'width' | 'height' | 'aspectRatio' | 'objectFit' | 'layout'>): Record<
+  string,
+  string | undefined
+> => {
+  const objectFitStyle = { 'object-fit': objectFit };
+  switch (layout) {
+    case 'fixed':
+      return {
+        ...objectFitStyle,
+        width: isValid(width) ? `${width}px` : undefined,
+        height: isValid(height) ? `${height}px` : undefined,
+      };
+    case 'constrained':
+      return {
+        ...objectFitStyle,
+        width: '100%',
+        'max-width': isValid(width) ? `${width}px` : undefined,
+        'max-height': isValid(height) ? `${height}px` : undefined,
+        'aspect-ratio': isValid(aspectRatio) ? `${aspectRatio}` : undefined,
```

Comment:
IMHO this is the only thing allowed to vary (with 3 digits max) and it should result in a few classes to inject